lib/Logger/LogContext.cpp

```diff
@@ -0,0 +1,122 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2021 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Manuel Pöter
+////////////////////////////////////////////////////////////////////////////////
+
+#include "LogContext.h"
+
+#include "Basics/debugging.h"
+
+#include <string>
+#include <vector>
+
+using namespace arangodb;
+
+namespace {
+  thread_local LogContext localLogContext;
+}
+
+struct LogContext::Impl {
+  std::vector<std::pair<std::string, std::string>> values;
+};
+
+LogContext::ScopedValue::ScopedValue(std::string key, std::string value) {
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+  _newValue = value;
+#endif
+  std::tie(_idx, _oldValue) = LogContext::current().set(std::move(key), std::move(value));
+}
+
+LogContext::ScopedValue::~ScopedValue() {
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+  TRI_ASSERT(_idx < LogContext::current()._impl->values.size());
+  TRI_ASSERT(LogContext::current()._impl->values[_idx].second == _newValue);
+#endif
+  LogContext::current().restore(_idx, std::move(_oldValue));
```

Comment:
Maybe we should not rely on LogContext::current()  is the same as it was provided in the ScopedValue constructor?
E.g. One could create a ScopedValue as class member. Set some "context"  at class instance creation. Then that instance could be transferred to another thread  and dtor will be called by some another thread. So we will set values in one LogContext and restore to some another LogContext.

Reply:
A `ScopedValue` should not be stored in a data member of some data structure that is passed between different threads. `ScopedValue`, `ScopedContext` etc. should only ever be created on the stack. Unfortunately there is not good way to enforce this in C++. I can (and should) make `ScopedValue` non-copyable though.

Reply:
Maybe at least store thread_id in the ScopedValue ctor and ASSERT it on dtor ?

Reply:
Good idea, will do!